File: examples/lh-server/go/08-exception-handling/workflow.go
package main

import "github.com/littlehorse-enterprises/littlehorse/sdk-go/littlehorse"

func Wflogic(wf *littlehorse.WorkflowThread) {
	price := wf.DeclareDouble("price").Required()
	itemID := wf.DeclareStr("item").WithDefault("lightsaber")
	userID := wf.DeclareStr("user-id").WithDefault("obiwan")
	charge := wf.Execute("charge-credit-card", userID, price)

	insufficientFunds := "insufficient-funds"
	wf.HandleException(charge, &insufficientFunds, func(handler *littlehorse.WorkflowThread) {
		handler.Execute("cancel-order-insufficient-funds", userID)
		message := "Credit card did not have sufficient funds"
		handler.Fail(nil, "insufficient-funds", &message)
	})
	wf.HandleError(charge, nil, func(handler *littlehorse.WorkflowThread) {
		handler.Execute("notify-order-failed", userID)
		message := "Failed to charge credit card"
		handler.Fail(nil, "technical-failure", &message)
	})
	wf.Execute("ship-item", itemID, userID)
}

File: examples/lh-server/python/08-exception-handling/workflow.py
from littlehorse.workflow import Workflow, WorkflowThread


def get_workflow() -> Workflow:
    def wfLogic(wf: WorkflowThread) -> None:
        price = wf.declare_double("price").required()
        item_id = wf.declare_str("item")
        item_id.assign("lightsaber")
        user_id = wf.declare_str("user-id")
        user_id.assign("obiwan")
        charge_credit_card_handle = wf.execute("charge-credit-card", user_id, price)

        def insufficient_funds(handler: WorkflowThread) -> None:
            handler.execute("cancel-order-insufficient-funds", user_id)
            handler.fail(
                "insufficient-funds",
                "Credit card did not have sufficient funds",
            )

        def technical_failure(handler: WorkflowThread) -> None:
            handler.execute("notify-order-failed", user_id)
            handler.fail("technical-failure", "Failed to charge credit card")

        wf.handle_exception(
            charge_credit_card_handle,
            insufficient_funds,
            "insufficient-funds",
        )
        wf.handle_error(charge_credit_card_handle, technical_failure)
        wf.execute("ship-item", item_id, user_id)

    return Workflow("exception-example", wfLogic)
